### PR TITLE
Rewrite some Future structs as async functions

### PR DESCRIPTION
**Pull request by expenses**

### CODE REVIEW

File: network/src/tests/validation.rs [hunk 1]
```diff
@@ -66,77 +66,48 @@ fn clone_gossip(n: &TopicNotification) -> TopicNotification {
...
-			}
+async fn gossip_router(
+	mut incoming_messages: mpsc::UnboundedReceiver<(Hash, TopicNotification)>,
+	mut incoming_streams: mpsc::UnboundedReceiver<(Hash, mpsc::UnboundedSender<TopicNotification>)>
+) {
+	let mut outgoing: Vec<(Hash, mpsc::UnboundedSender<TopicNotification>)> = Vec::new();
+	let mut messages = Vec::new();
+
+	loop {
+		match select(incoming_messages.next(), incoming_streams.next()).await {
```
Comment (montekki) [inline network/src/tests/validation.rs]: Why not use the [`futures::select`](https://docs.rs/futures/0.3.1/futures/macro.select.html)?

Comment (expenses) [reply network/src/tests/validation.rs]: I prefer to avoid macros where possible.

**Comment by tomaka** [comment]
Would it be possible to merge #680 beforehand? I'm becoming crazy with merge conflicts again.

**Comment by expenses** [comment]
@tomaka of course!

**Comment by montekki** [comment]
Running two nodes panics:


```
2019-12-11 16:36:11 Discovered new external address for our node: /ip4/127.0.0.1/tcp/30333/p2p/QmS5D1uFXjME8oC3vYRmN9sWCuMYJJNCnRqN7CrXi7xbPi
2019-12-11 16:36:12 Starting parachain attestation session on top of parent 0x21614ca6680f7dc9794cb66593603a545e5ab0d03dc3dbf434258a7e51156475. Local parachain duty is Some(LocalDuty { validation: Relay, index: 1 })

====================

Version: 0.7.9-4c4483b8-x86_64-linux-gnu

stack backtrace:
   0: sp_panic_handler::set::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   2: std::panicking::begin_panic
   3: tokio::runtime::blocking::pool::spawn_blocking
   4: <std::future::GenFuture<T> as core::future::future::Future>::poll
   5: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
   6: <futures_util::future::try_future::map_ok::MapOk<Fut,F> as core::future::future::Future>::poll
   7: futures_util::future::try_future::try_chain::TryChain<Fut1,Fut2,Data>::poll
   8: <futures_util::future::future::then::Then<Fut1,Fut2,F> as core::future::future::Future>::poll
   9: <futures_util::compat::compat03as01::Compat<Fut> as futures::future::Future>::poll
  10: std::panicking::try::do_call
  11: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  12: <futures::future::catch_unwind::CatchUnwind<F> as futures::future::Future>::poll
  13: futures::future::chain::Chain<A,B,C>::poll
  14: <futures::future::select::Select<A,B> as futures::future::Future>::poll
  15: futures::future::chain::Chain<A,B,C>::poll
  16: futures::task_impl::std::set
  17: std::panicking::try::do_call
  18: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  19: tokio_threadpool::task::Task::run
  20: tokio_threadpool::worker::Worker::run_task
  21: tokio_threadpool::worker::Worker::run
  22: std::thread::local::LocalKey<T>::with
  23: std::thread::local::LocalKey<T>::with
  24: std::thread::local::LocalKey<T>::with
  25: tokio::runtime::threadpool::builder::Builder::build::{{closure}}
  26: std::thread::local::LocalKey<T>::with
  27: std::thread::local::LocalKey<T>::with
  28: std::sys_common::backtrace::__rust_begin_short_backtrace
  29: std::panicking::try::do_call
  30: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  31: core::ops::function::FnOnce::call_once{{vtable.shim}}
  32: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/4560ea788cb760f0a34127156c78e2552949f734/src/liballoc/boxed.rs:922
  33: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/4560ea788cb760f0a34127156c78e2552949f734/src/liballoc/boxed.rs:922
      std::sys_common::thread::start_thread
             at src/libstd/sys_common/thread.rs:13
      std::sys::unix::thread::Thread::new::thread_start
             at src/libstd/sys/unix/thread.rs:79
  34: start_thread
  35: clone


Thread 'tokio-runtime-worker-2' panicked at 'not currently running on the Tokio runtime.', /home/theodor/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.4/src/runtime/blocking/pool.rs:86

This is a bug. Please report it at:

        https://github.com/paritytech/polkadot/issues/new

```

Way to reproduce:

```
 cargo run --release -- --chain=local --validator --alice  -d /tmp/alice

```

```
 cargo run --release -- --bob --chain local -d /tmp/bob/ --validator --port 30336 --bootnodes --bootnodes='/ip4/127.0.0.1/tcp/30333/p2p/$ALICE_NODE_ID'

```

**Comment by expenses** [comment]
> Running two nodes panics:
> 
> ```
> 2019-12-11 16:36:11 Discovered new external address for our node: /ip4/127.0.0.1/tcp/30333/p2p/QmS5D1uFXjME8oC3vYRmN9sWCuMYJJNCnRqN7CrXi7xbPi
> 2019-12-11 16:36:12 Starting parachain attestation session on top of parent 0x21614ca6680f7dc9794cb66593603a545e5ab0d03dc3dbf434258a7e51156475. Local parachain duty is Some(LocalDuty { validation: Relay, index: 1 })
> 
> ====================
> 
> Version: 0.7.9-4c4483b8-x86_64-linux-gnu
> 
> stack backtrace:
>    0: sp_panic_handler::set::{{closure}}
>    1: std::panicking::rust_panic_with_hook
>              at src/libstd/panicking.rs:477
>    2: std::panicking::begin_panic
>    3: tokio::runtime::blocking::pool::spawn_blocking
>    4: <std::future::GenFuture<T> as core::future::future::Future>::poll
>    5: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
>    6: <futures_util::future::try_future::map_ok::MapOk<Fut,F> as core::future::future::Future>::poll
>    7: futures_util::future::try_future::try_chain::TryChain<Fut1,Fut2,Data>::poll
>    8: <futures_util::future::future::then::Then<Fut1,Fut2,F> as core::future::future::Future>::poll
>    9: <futures_util::compat::compat03as01::Compat<Fut> as futures::future::Future>::poll
>   10: std::panicking::try::do_call
>   11: __rust_maybe_catch_panic
>              at src/libpanic_unwind/lib.rs:80
>   12: <futures::future::catch_unwind::CatchUnwind<F> as futures::future::Future>::poll
>   13: futures::future::chain::Chain<A,B,C>::poll
>   14: <futures::future::select::Select<A,B> as futures::future::Future>::poll
>   15: futures::future::chain::Chain<A,B,C>::poll
>   16: futures::task_impl::std::set
>   17: std::panicking::try::do_call
>   18: __rust_maybe_catch_panic
>              at src/libpanic_unwind/lib.rs:80
>   19: tokio_threadpool::task::Task::run
>   20: tokio_threadpool::worker::Worker::run_task
>   21: tokio_threadpool::worker::Worker::run
>   22: std::thread::local::LocalKey<T>::with
>   23: std::thread::local::LocalKey<T>::with
>   24: std::thread::local::LocalKey<T>::with
>   25: tokio::runtime::threadpool::builder::Builder::build::{{closure}}
>   26: std::thread::local::LocalKey<T>::with
>   27: std::thread::local::LocalKey<T>::with
>   28: std::sys_common::backtrace::__rust_begin_short_backtrace
>   29: std::panicking::try::do_call
>   30: __rust_maybe_catch_panic
>              at src/libpanic_unwind/lib.rs:80
>   31: core::ops::function::FnOnce::call_once{{vtable.shim}}
>   32: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
>              at /rustc/4560ea788cb760f0a34127156c78e2552949f734/src/liballoc/boxed.rs:922
>   33: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
>              at /rustc/4560ea788cb760f0a34127156c78e2552949f734/src/liballoc/boxed.rs:922
>       std::sys_common::thread::start_thread
>              at src/libstd/sys_common/thread.rs:13
>       std::sys::unix::thread::Thread::new::thread_start
>              at src/libstd/sys/unix/thread.rs:79
>   34: start_thread
>   35: clone
> 
> 
> Thread 'tokio-runtime-worker-2' panicked at 'not currently running on the Tokio runtime.', /home/theodor/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.4/src/runtime/blocking/pool.rs:86
> 
> This is a bug. Please report it at:
> 
>         https://github.com/paritytech/polkadot/issues/new
> ```
> 
> Way to reproduce:
> 
> ```
>  cargo run --release -- --chain=local --validator --alice  -d /tmp/alice
> ```
> 
> ```
>  cargo run --release -- --bob --chain local -d /tmp/bob/ --validator --port 30336 --bootnodes --bootnodes='/ip4/127.0.0.1/tcp/30333/p2p/$ALICE_NODE_ID'
> ```

That was intruduced here: https://github.com/paritytech/polkadot/pull/673/files#diff-79011ba421dbe9726f271e831317c90bL902-R905. This only happens on this PR, and not master, correct?

**Comment by expenses** [comment]
> > Running two nodes panics:
> > ```
> > 2019-12-11 16:36:11 Discovered new external address for our node: /ip4/127.0.0.1/tcp/30333/p2p/QmS5D1uFXjME8oC3vYRmN9sWCuMYJJNCnRqN7CrXi7xbPi
> > 2019-12-11 16:36:12 Starting parachain attestation session on top of parent 0x21614ca6680f7dc9794cb66593603a545e5ab0d03dc3dbf434258a7e51156475. Local parachain duty is Some(LocalDuty { validation: Relay, index: 1 })
> > 
> > ====================
> > 
> > Version: 0.7.9-4c4483b8-x86_64-linux-gnu
> > 
> > stack backtrace:
> >    0: sp_panic_handler::set::{{closure}}
> >    1: std::panicking::rust_panic_with_hook
> >              at src/libstd/panicking.rs:477
> >    2: std::panicking::begin_panic
> >    3: tokio::runtime::blocking::pool::spawn_blocking
> >    4: <std::future::GenFuture<T> as core::future::future::Future>::poll
> >    5: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
> >    6: <futures_util::future::try_future::map_ok::MapOk<Fut,F> as core::future::future::Future>::poll
> >    7: futures_util::future::try_future::try_chain::TryChain<Fut1,Fut2,Data>::poll
> >    8: <futures_util::future::future::then::Then<Fut1,Fut2,F> as core::future::future::Future>::poll
> >    9: <futures_util::compat::compat03as01::Compat<Fut> as futures::future::Future>::poll
> >   10: std::panicking::try::do_call
> >   11: __rust_maybe_catch_panic
> >              at src/libpanic_unwind/lib.rs:80
> >   12: <futures::future::catch_unwind::CatchUnwind<F> as futures::future::Future>::poll
> >   13: futures::future::chain::Chain<A,B,C>::poll
> >   14: <futures::future::select::Select<A,B> as futures::future::Future>::poll
> >   15: futures::future::chain::Chain<A,B,C>::poll
> >   16: futures::task_impl::std::set
> >   17: std::panicking::try::do_call
> >   18: __rust_maybe_catch_panic
> >              at src/libpanic_unwind/lib.rs:80
> >   19: tokio_threadpool::task::Task::run
> >   20: tokio_threadpool::worker::Worker::run_task
> >   21: tokio_threadpool::worker::Worker::run
> >   22: std::thread::local::LocalKey<T>::with
> >   23: std::thread::local::LocalKey<T>::with
> >   24: std::thread::local::LocalKey<T>::with
> >   25: tokio::runtime::threadpool::builder::Builder::build::{{closure}}
> >   26: std::thread::local::LocalKey<T>::with
> >   27: std::thread::local::LocalKey<T>::with
> >   28: std::sys_common::backtrace::__rust_begin_short_backtrace
> >   29: std::panicking::try::do_call
> >   30: __rust_maybe_catch_panic
> >              at src/libpanic_unwind/lib.rs:80
> >   31: core::ops::function::FnOnce::call_once{{vtable.shim}}
> >   32: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
> >              at /rustc/4560ea788cb760f0a34127156c78e2552949f734/src/liballoc/boxed.rs:922
> >   33: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
> >              at /rustc/4560ea788cb760f0a34127156c78e2552949f734/src/liballoc/boxed.rs:922
> >       std::sys_common::thread::start_thread
> >              at src/libstd/sys_common/thread.rs:13
> >       std::sys::unix::thread::Thread::new::thread_start
> >              at src/libstd/sys/unix/thread.rs:79
> >   34: start_thread
> >   35: clone
> > 
> > 
> > Thread 'tokio-runtime-worker-2' panicked at 'not currently running on the Tokio runtime.', /home/theodor/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.4/src/runtime/blocking/pool.rs:86
> > 
> > This is a bug. Please report it at:
> > 
> >         https://github.com/paritytech/polkadot/issues/new
> > ```
> > 
> > 
> > Way to reproduce:
> > ```
> >  cargo run --release -- --chain=local --validator --alice  -d /tmp/alice
> > ```
> > 
> > 
> > ```
> >  cargo run --release -- --bob --chain local -d /tmp/bob/ --validator --port 30336 --bootnodes --bootnodes='/ip4/127.0.0.1/tcp/30333/p2p/$ALICE_NODE_ID'
> > ```
> 
> That was introduced here: [938f411](https://github.com/paritytech/polkadot/commit/938f411a9365e9c5fb16bfedb62aacac4403d063) as I believed that the panic did not occur. This only happens on this PR, and not master, correct?

Ok, it does happen on master. Shame we didn't catch that. I'll get on a PR.

**Comment by montekki** [comment]
> Ok, it does happen on master. Shame we didn't catch that. 

Yep, looks like it

File: network/src/tests/validation.rs [hunk 2]
```diff
@@ -66,77 +66,48 @@ fn clone_gossip(n: &TopicNotification) -> TopicNotification {
...
+				outgoing.retain(|&(ref o_topic, ref sender)| {
+					o_topic != &topic || sender.unbounded_send(clone_gossip(&message)).is_ok()
+				});
+				messages.push((topic, message));
+			},
+			Either::Right((Some((topic, sender)), _)) => {
+				for message in messages.iter()
+					.filter(|&&(ref t, _)| t == &topic)
+					.map(|&(_, ref msg)| clone_gossip(msg))
+					{
```
Comment (gavofyork) [inline network/src/tests/validation.rs]: One indent level too much (should be anchored to the `for` level)

**Comment by expenses** [comment]
@montekki Thanks! That should be fixed now.

**Comment by expenses** [comment]
@tomaka Do you still want me to hold off on merging this?

**Comment by montekki** [comment]
@expenses yes, now it's working.

**Pull request merged**